File: testdata/precision-suite-cpp/tp_cmdinjection.c
// Command injection: an environment variable (standing in for any untrusted
// input) is concatenated into a shell command with strcat and handed to
// system(3), which runs it through /bin/sh -c. There is no allow-list and no
// escaping, so a value like `; rm -rf /` executes. This is CWE-78. A correct
// scanner fires TAINT-002; nox's C/C++ taint model matches the tainted value
// flowing into the system() call.
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void generate_report(void) {
    char *name = getenv("REPORT");
    char cmd[256];
    strcpy(cmd, "generate-report ");
    strcat(cmd, name);
    system(cmd); // nox-expect: TAINT-002
}

File: testdata/precision-suite-dart/tp_cmdinjection.dart
// Command injection: an attacker-controlled value (an environment variable
// standing in for any untrusted input) is string-interpolated into a shell
// command and handed to `Process.run` with `runInShell: true`, so /bin/sh -c
// executes the interpolated string. This is CWE-78. A correct scanner fires
// TAINT-002. nox's Dart taint model sees the tainted value reach the `$name`
// interpolation (which lexctx classifies as CODE) and then the `Process.run`
// sink with no allow-list.
import 'dart:io';

Future<void> runReport() async {
  final name = Platform.environment['REPORT'] ?? '';
  await Process.run('sh', ['-c', 'generate-report $name'], runInShell: true); // nox-expect: TAINT-002
}
